File: WorldSystems/MarkersLabels/MarkersAndLabels.cs
using CloudRP.WorldSystems.MarkersLabels;
using GTANetworkAPI;
using Newtonsoft.Json;
using System;
using System.Collections.Generic;
using System.Text;

namespace CloudRP.World.MarkersLabels
{
    public class MarkersAndLabels
    {
        public static TextLabel setTextLabel(Vector3 position, string text, float range, uint dim = 0)
        {
            return NAPI.TextLabel.CreateTextLabel(text, position, range, 1.0f, 4, new Color(255, 255, 255, 255), true, dim);
        }

        public static Marker setPlaceMarker(Vector3 position, uint dim = 0)
        {
            return NAPI.Marker.CreateMarker(27, new Vector3(position.X, position.Y, position.Z - 0.92), new Vector3(0, 0, 0), new Vector3(0, 0, 0), 0.5f, new Color(214, 175, 250, 250), false, dim);
        }

        public static void addBlipForClient(Player target, int blipSprite, string name, Vector3 pos, int colour, int alpha = 255, int timeout = 60, bool setRoute = false, bool setMarker = false)
        {
            removeClientBlip(target);
            target.TriggerEvent("clientBlip:addClientBlip", blipSprite, name, pos, colour, alpha, timeout, setRoute, setMarker);
        }

        public static void removeClientBlip(Player player)
        {
            player.TriggerEvent("clientBlip:removeClientBlip");
        }

        public static void setClientWaypoint(Player player, Vector3 pos)
            => player.TriggerEvent("clientBlip:setWaypoint", pos);

        public static void loadClientBlips(Player player, List<Vector3> targetBlips, string name, int blipType, int blipColour, bool setMarker)
        {
            List<ClientBlip> blips = new List<ClientBlip>();

            targetBlips.ForEach(blip =>
            {
                blips.Add(new ClientBlip
                {
                    blipId = targetBlips.IndexOf(blip),
                    colour = blipColour,
                    name = name,
                    pos = blip,
                    setMarker = setMarker
                });
            });

            player.TriggerEvent("clientBlip:addArrayOfBlips", blips);
        }

        public static void deleteClientBlip(Player player, int blipId)
            => player.TriggerEvent("clientBlip:handleBlipDelete", blipId);

        public static void flushClientBlips(Player player)
            => player.TriggerEvent("clientBlip:flushBlipArray");
    }
}

File: ClientSide/client/src/client/MarkersAndLabels/MarkersAndLabels.ts
import { ClientBlip } from "@/@types";

export default class MarkersAndLabels {
    public static LocalPlayer: PlayerMp;
    public static readonly defaultBlipTimeout_seconds: number = 30;
    private static createBlip: BlipMp | undefined;
    private static createMarker: MarkerMp | undefined;
    private static deleteInterval: ReturnType<typeof setInterval> | undefined;
    private static blips: Map<number, ClientBlip> = new Map<number, ClientBlip>();

    constructor() {
        MarkersAndLabels.LocalPlayer = mp.players.local;

        mp.events.add("render", MarkersAndLabels.removeNorthBlip);

        mp.events.add("clientBlip:addClientBlip", MarkersAndLabels.addClientBlip);
        mp.events.add("clientBlip:removeClientBlip", MarkersAndLabels.removeClientBlip);
        mp.events.add("clientBlip:setWaypoint", MarkersAndLabels.setWaypoint);
        mp.events.add("clientBlip:addArrayOfBlips", MarkersAndLabels.addArrayOfBlips);
        mp.events.add("clientBlip:handleBlipDelete", MarkersAndLabels.deleteFromBlipArray);
        mp.events.add("clientBlip:flushBlipArray", MarkersAndLabels.flushBlipArray);
    }

    private static addArrayOfBlips(blips: ClientBlip[]) {
        blips.forEach(blip => {
            blip.blip = mp.blips.new(blip.type, blip.pos, {
                name: blip.name,
                color: blip.colour
            });

            if (blip.setMarker) {
                blip.marker = mp.markers.new(1, new mp.Vector3(blip.pos.x, blip.pos.y, blip.pos.z - 1), 2, {
                    color: [0, 255, 0, 30],
                    dimension: 0
                });
            }

            MarkersAndLabels.blips.set(blip.blipId, blip);
        });
    }

    private static deleteFromBlipArray(blipId: number) {
        let blip: ClientBlip | undefined = MarkersAndLabels.blips.get(blipId);

        if (!blip) return;

        if (mp.blips.exists(blip.blip)) {
            blip.blip.destroy();

            if(blip.setMarker && mp.markers.exists(blip.marker)) blip.marker.destroy();

            MarkersAndLabels.blips.delete(blipId);
        }
    }

    private static flushBlipArray() {
        MarkersAndLabels.blips.forEach(blip => {
            if (blip.blip && mp.blips.exists(blip.blip)) {
                blip.blip.destroy();

                if(blip.setMarker && mp.markers.exists(blip.marker)) blip.marker.destroy(); 

                MarkersAndLabels.blips.delete(blip.blipId);
            }
        });

    }

    public static setWaypoint(coords: Vector3) {
        mp.game.ui.setNewWaypoint(coords.x, coords.y);
    }

    public static removeNorthBlip() {
        mp.game.ui.setBlipAlpha(mp.game.invoke("0x3F0CF9CB7E589B88"), 0);
    }

    public static addClientBlip(blipSprite: number, name: string, position: Vector3, blipColour: number, alpha: number = 255, timeout: number = MarkersAndLabels.defaultBlipTimeout_seconds, setRoute: boolean = false, setMarker: boolean = false) {
        MarkersAndLabels.removeClientBlip();

        let createdBlip: BlipMp | null = mp.blips.new(blipSprite, position, {
            alpha: alpha,
            color: blipColour,
            dimension: 0,
            drawDistance: 1.0,
            name: name,
            rotation: 0,
            scale: 1.0,
            shortRange: false
        });

        MarkersAndLabels.createBlip = createdBlip;

        createdBlip.setRoute(setRoute);

        if (setMarker) {
            MarkersAndLabels.createMarker = mp.markers.new(1, new mp.Vector3(position.x, position.y, position.z - 1), 2, {
                color: [0, 255, 0, 30],
                dimension: 0
            });
        }

        if (timeout != -1) {
            MarkersAndLabels.deleteInterval = setTimeout(() => {
                MarkersAndLabels.removeClientBlip();
            }, timeout * 1000);
        }
    }

    public static removeClientBlip() {
        if (MarkersAndLabels.deleteInterval) {
            clearInterval(MarkersAndLabels.deleteInterval);
            MarkersAndLabels.deleteInterval = undefined;
        }

        if (MarkersAndLabels.createMarker) {
            MarkersAndLabels.createMarker.destroy();
            MarkersAndLabels.createMarker = undefined;
        }

        if (MarkersAndLabels.createBlip) {
            MarkersAndLabels.createBlip.destroy();
            MarkersAndLabels.createBlip = undefined;
        }
    }
}